beacon_node/lighthouse_network/src/peer_manager/connectivity.rs

```diff
@@ -0,0 +1,188 @@
+use crate::peer_manager::network_globals_provider::NetworkGlobalsProvider;
+use crate::peer_manager::PeerManagerEvent;
+use crate::rpc::GoodbyeReason;
+use crate::{EnrExt, PeerId};
+use discv5::Enr;
+use smallvec::SmallVec;
+use std::collections::HashMap;
+use std::time::Instant;
+
+pub struct Connectivity<N: NetworkGlobalsProvider> {
+    target_peers: usize,
+    peer_excess_factor: f32,
+    priority_peer_excess: f32,
+    min_outbound_only_factor: f32,
+    target_outbound_only_factor: f32,
+    /// Keeps track of whether the discovery service is enabled or not.
+    discovery_enabled: bool,
+    /// Peers queued to be dialed.
+    peers_to_dial: Vec<Enr>,
+    network_globals_provider: N,
+}
+
+impl<N: NetworkGlobalsProvider> Connectivity<N> {
+    pub fn new(
+        target_peers: usize,
+        peer_excess_factor: f32,
+        priority_peer_excess: f32,
+        min_outbound_only_factor: f32,
+        target_outbound_only_factor: f32,
+        discovery_enabled: bool,
+        network_globals_provider: N,
+    ) -> Self {
+        Self {
+            target_peers,
+            peer_excess_factor,
+            priority_peer_excess,
+            min_outbound_only_factor,
+            target_outbound_only_factor,
+            discovery_enabled,
+            peers_to_dial: Default::default(),
+            network_globals_provider,
+        }
+    }
+
+    /// A peer is being dialed.
+    /// Returns true, if this peer will be dialed.
+    pub fn dial_peer(&mut self, peer: Enr) -> bool {
+        if self.network_globals_provider.should_dial(&peer.peer_id()) {
+            self.peers_to_dial.push(peer);
+            true
+        } else {
+            false
+        }
+    }
+
+    // Gracefully disconnects a peer without banning them.
+    pub fn disconnect_peer(
+        &mut self,
+        peer_id: PeerId,
+        reason: GoodbyeReason,
+        events: &mut SmallVec<[PeerManagerEvent; 16]>,
+    ) {
+        events.push(PeerManagerEvent::DisconnectPeer(peer_id, reason));
+        self.network_globals_provider
+            .notify_disconnecting(&peer_id, false);
+    }
+
+    pub fn next_peer_to_dial(&mut self) -> Option<Enr> {
+        self.peers_to_dial.pop()
+    }
+
+    /// Peers that have been returned by discovery requests that are suitable for dialing are
+    /// returned here.
+    ///
+    /// This function decides whether or not to dial these peers.
```

Comment:
These comments should describe what the return value is used for as this logic has changed here and it might be easier for future people to understand when looking at what these functions do